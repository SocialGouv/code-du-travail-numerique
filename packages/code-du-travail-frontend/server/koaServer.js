const Koa = require("koa");
const bodyParser = require("koa-bodyparser");
const helmet = require("koa-helmet");
const Router = require("koa-router");
const redirects = require("./redirects.json");

const IS_PRODUCTION_DEPLOYMENT =
  process.env.NEXT_PUBLIC_IS_PRODUCTION_DEPLOYMENT === "true";
const PROD_HOSTNAME = process.env.PROD_HOSTNAME || "code.travail.gouv.fr";
const AZURE_BASE_URL =
  process.env.AZURE_BASE_URL || "https://cdtnadmindev.blob.core.windows.net";

const dev = process.env.NODE_ENV !== "production";

const robotsDev = ["User-agent: *", "Disallow: /"].join("\n");
const robotsProd = [
  "User-agent: *",
  "Disallow: /assets/",
  "Disallow: /images/",
  "",
  `Sitemap: https://${PROD_HOSTNAME}/sitemap.xml`,
].join("\n");

async function getKoaServer({ nextApp }) {
  const nextHandler = nextApp.getRequestHandler();
  const server = new Koa();
  const router = new Router();

  const cspConfig = {
    directives: {
      defaultSrc: [
        "'self'",
        "*.travail.gouv.fr",
        "*.data.gouv.fr",
        "*.fabrique.social.gouv.fr",
      ],
      "font-src": ["'self'", "data:", "blob:"],
      frameSrc: [
        "'self'",
        "https://mon-entreprise.fr",
        "https://matomo.fabrique.social.gouv.fr",
        "*.dailymotion.com",
        "*.doubleclick.net",
      ],
      imgSrc: [
        "'self'",
        "data:",
        "*.fabrique.social.gouv.fr",
        "https://travail-emploi.gouv.fr",
        "https://mon-entreprise.fr",
        AZURE_BASE_URL,
        "*.doubleclick.net",
        "*.xiti.com",
      ],
      scriptSrc: [
        "'self'",
        "https://mon-entreprise.fr",
        "*.fabrique.social.gouv.fr",
        "https://cdnjs.cloudflare.com",
        "https://www.googletagmanager.com",
        "'unsafe-inline'",
        "*.doubleclick.net",
      ],
      styleSrc: ["'self'", "'unsafe-inline'"],
      ...(process.env.NEXT_PUBLIC_SENTRY_DSN && {
        reportUri: process.env.NEXT_PUBLIC_SENTRY_DSN,
      }),
    },
  };
  if (dev) {
    server.use(bodyParser());
    cspConfig.directives.scriptSrc.push("'unsafe-eval'");
  }
  server.use(helmet.contentSecurityPolicy(cspConfig));

  if (!IS_PRODUCTION_DEPLOYMENT) {
    server.use(async function (ctx, next) {
      ctx.set({ "X-Robots-Tag": "noindex, nofollow, nosnippet" });
      await next();
    });
  }

  router.get("/health", async (ctx) => {
    ctx.body = { status: "up and running" };
  });

  router.get("/robots.txt", async (ctx) => {
    ctx.type = "text/plain";
    ctx.body = IS_PRODUCTION_DEPLOYMENT ? robotsProd : robotsDev;
  });

  redirects.forEach(({ baseUrl, code, redirectUrl }) => {
    router.get(baseUrl, async (ctx) => {
      ctx.status = code;
      ctx.set("Location", redirectUrl);
      ctx.redirect(redirectUrl);
    });
  });

  router.all("(.*)", async (ctx) => {
    await nextHandler(ctx.req, ctx.res);
    ctx.respond = false;
  });

  server.use(async (ctx, next) => {
    ctx.status = 200;
    await next();
  });

  server.use(router.routes());
  return server;
}

module.exports = { getKoaServer };
