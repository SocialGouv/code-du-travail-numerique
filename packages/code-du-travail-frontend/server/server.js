// Import prerequisite packages
const next = require("next");
const { getKoaServer } = require("./koaServer");
/**
 * this env variable is use to target developpement / staging deployement
 * in order to block indexing bot using a x-robot-header and an appropriate robots.txt
 */
const IS_PRODUCTION_DEPLOYMENT =
  process.env.NEXT_PUBLIC_IS_PRODUCTION_DEPLOYMENT === "true";
const PORT = parseInt(process.env.FRONTEND_PORT, 10) || 3000;
const FRONTEND_HOST = process.env.FRONTEND_HOST || `http://localhost:${PORT}`;
const PROD_HOSTNAME = process.env.PROD_HOSTNAME || "code.travail.gouv.fr";
const NEXT_PUBLIC_SENTRY_DSN = process.env.NEXT_PUBLIC_SENTRY_DSN || "";
const PACKAGE_VERSION = process.env.VERSION || "";

const dev = process.env.NODE_ENV !== "production";

// Initialize NextJs instance and expose request handler
const nextApp = next({ dev });

nextApp.prepare().then(async () => {
  const server = await getKoaServer({
    nextApp,
  });

  server.listen(PORT, () => {
    //eslint-disable-next-line no-console
    console.log(`
  â€º Ready on ${FRONTEND_HOST}

  Environment:

  - process.env.NODE_ENV : ${process.env.NODE_ENV || "development"}
  - version: ${PACKAGE_VERSION}
  - is deployed on prod(${PROD_HOSTNAME}): ${IS_PRODUCTION_DEPLOYMENT}
  - sentry: ${NEXT_PUBLIC_SENTRY_DSN},

`);
  });

  this.server = server;
});

module.exports = nextApp;
