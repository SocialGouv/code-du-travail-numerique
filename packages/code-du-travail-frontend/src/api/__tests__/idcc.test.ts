import request from "supertest";
import server from "nextjs-http-supertest";

describe("IDCC", () => {
  afterAll(() => {
    server.close();
  });

  it("should return idcc for banques", async () => {
    const res = await request(server).get("/api/idcc?q=banque");
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });

  it("should return nothing", async () => {
    const res = await request(server).get("/api/idcc?q=blabla");
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });
});
