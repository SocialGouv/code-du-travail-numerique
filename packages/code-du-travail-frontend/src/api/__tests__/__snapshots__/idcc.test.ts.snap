// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IDCC return idcc for number 27 1`] = `
{
  "hits": [
    {
      "_id": "20",
      "_index": "cdtn_test-v2_documents",
      "_score": 7.067604,
      "_source": {
        "cdtnId": "ceb38cca3c",
        "contributions": false,
        "effectif": 219,
        "id": "KALICONT000023973390",
        "num": 2701,
        "shortTitle": "Banque : personnel des banques de la Guyane",
        "slug": "2701-banque-personnel-des-banques-de-la-guyane",
        "title": "Banque : personnel des banques de la Guyane",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000023973390",
      },
      "_type": "_doc",
    },
  ],
  "max_score": 7.067604,
  "total": {
    "relation": "eq",
    "value": 1,
  },
}
`;

exports[`IDCC return idcc results for boulangerie 1`] = `
{
  "hits": [
    {
      "_id": "19",
      "_index": "cdtn_test-v2_documents",
      "_score": 15.546368,
      "_source": {
        "cdtnId": "647c224c9b",
        "contributions": true,
        "effectif": 134889,
        "id": "KALICONT000005635886",
        "num": 843,
        "shortTitle": "Boulangerie-pâtisserie (entreprises artisanales)",
        "slug": "843-boulangerie-patisserie-entreprises-artisanales",
        "title": "Boulangerie-pâtisserie (entreprises artisanales)",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635886",
      },
      "_type": "_doc",
    },
    {
      "_id": "18",
      "_index": "cdtn_test-v2_documents",
      "_score": 14.4430065,
      "_source": {
        "cdtnId": "16752da066",
        "contributions": false,
        "effectif": 44699,
        "id": "KALICONT000005635691",
        "num": 1747,
        "shortTitle": "Activités industrielles de boulangerie et pâtisserie",
        "slug": "1747-activites-industrielles-de-boulangerie-et-patisserie",
        "title": "Activités industrielles de boulangerie et pâtisserie",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635691",
      },
      "_type": "_doc",
    },
  ],
  "max_score": 15.546368,
  "total": {
    "relation": "eq",
    "value": 2,
  },
}
`;

exports[`IDCC return idcc results for pati 1`] = `
{
  "hits": [
    {
      "_id": "19",
      "_index": "cdtn_test-v2_documents",
      "_score": 15.258489,
      "_source": {
        "cdtnId": "647c224c9b",
        "contributions": true,
        "effectif": 134889,
        "id": "KALICONT000005635886",
        "num": 843,
        "shortTitle": "Boulangerie-pâtisserie (entreprises artisanales)",
        "slug": "843-boulangerie-patisserie-entreprises-artisanales",
        "title": "Boulangerie-pâtisserie (entreprises artisanales)",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635886",
      },
      "_type": "_doc",
    },
    {
      "_id": "18",
      "_index": "cdtn_test-v2_documents",
      "_score": 14.155128,
      "_source": {
        "cdtnId": "16752da066",
        "contributions": false,
        "effectif": 44699,
        "id": "KALICONT000005635691",
        "num": 1747,
        "shortTitle": "Activités industrielles de boulangerie et pâtisserie",
        "slug": "1747-activites-industrielles-de-boulangerie-et-patisserie",
        "title": "Activités industrielles de boulangerie et pâtisserie",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635691",
      },
      "_type": "_doc",
    },
  ],
  "max_score": 15.258489,
  "total": {
    "relation": "eq",
    "value": 2,
  },
}
`;

exports[`IDCC return idcc results in correct order for banque 1`] = `
{
  "hits": [
    {
      "_id": "21",
      "_index": "cdtn_test-v2_documents",
      "_score": 17.23316,
      "_source": {
        "cdtnId": "a25dfc974f",
        "contributions": true,
        "effectif": 216431,
        "id": "KALICONT000005635780",
        "num": 2120,
        "shortTitle": "Banque",
        "slug": "2120-banque",
        "title": "Banque",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635780",
      },
      "_type": "_doc",
    },
    {
      "_id": "20",
      "_index": "cdtn_test-v2_documents",
      "_score": 10.528515,
      "_source": {
        "cdtnId": "ceb38cca3c",
        "contributions": false,
        "effectif": 219,
        "id": "KALICONT000023973390",
        "num": 2701,
        "shortTitle": "Banque : personnel des banques de la Guyane",
        "slug": "2701-banque-personnel-des-banques-de-la-guyane",
        "title": "Banque : personnel des banques de la Guyane",
        "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000023973390",
      },
      "_type": "_doc",
    },
  ],
  "max_score": 17.23316,
  "total": {
    "relation": "eq",
    "value": 2,
  },
}
`;

exports[`IDCC should return nothing 1`] = `
{
  "_shards": {
    "failed": 0,
    "skipped": 0,
    "successful": 1,
    "total": 1,
  },
  "hits": {
    "hits": [],
    "max_score": null,
    "total": {
      "relation": "eq",
      "value": 0,
    },
  },
  "timed_out": false,
}
`;
